[Shapes] - Implement Rectangle widget (resolves #26)

suragch opened the pull request:
@matthew-carroll I'm going to create an initial PR now so that I can get feedback before continuing too far in a direction that we don't want to go.

This PR implements a minimal `Rectangle` widget in Flutter. Specifically you'll find the following additions to the repo:

- `Rectangle` examples in the SwiftUI gallery app. I also added Shapes to the More bottom nav bar.

<img width="250" alt="Screenshot 2024-06-06 at 12 04 12" src="https://github.com/Flutter-Bounty-Hunters/swift_ui/assets/10300220/be5a168d-d35d-4878-b9dd-33044b0bf6bf">
<img width="238" alt="Screenshot 2024-06-06 at 12 06 08" src="https://github.com/Flutter-Bounty-Hunters/swift_ui/assets/10300220/b6ab270c-168f-4ef1-a7c8-9c30a8f65ec0">
<img width="234" alt="Screenshot 2024-06-06 at 12 06 26" src="https://github.com/Flutter-Bounty-Hunters/swift_ui/assets/10300220/7ad49db6-05da-4466-bf56-d3bf7e7689a7">

- Mirrored the navigation in the Flutter demo app, but only implemented two rectangle styles so far:

<img width="321" alt="Screenshot 2024-06-06 at 12 07 50" src="https://github.com/Flutter-Bounty-Hunters/swift_ui/assets/10300220/99f5f53f-1079-43a7-af52-873e21aae896">
<img width="321" alt="Screenshot 2024-06-06 at 12 08 02" src="https://github.com/Flutter-Bounty-Hunters/swift_ui/assets/10300220/f51b646e-dbd1-455d-a3a8-53b768cdfe4a">
<img width="321" alt="Screenshot 2024-06-06 at 12 08 08" src="https://github.com/Flutter-Bounty-Hunters/swift_ui/assets/10300220/cba3bc6c-9d79-48c7-a8d3-c110cc5458db">

## Issues to consider

- In SwiftUI, the `fill` is a `ShapeStyle`, which can be a color, gradient, or other pattern. I'm not really sure what the best way is to implement this in Flutter. I created a ShapeStyle class but maybe it would be better to just have color and gradient be properties on Rectangle. Also, `ShapeStyle` in SwiftUI doesn't have a color property. I think it might get color from the ForegroundStyle. It doesn't have a `gradient` property either. Instead it has a `linearGradient`, `radialGradient`, etc. That didn't seem very Flutter-like to add all of those properties when they can all be expressed with `Gradient`.
- Should there be some more general `Shape` class that ties all of the shapes together, or should we implement each shape widget independently?
- I haven't added tests yet. I'll do that after I'm a little more sure of the direction. I also have to learn how to make golden tests.

I don't think squashing is currently enabled for this repo. Could you [enable it](https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/configuring-pull-request-merges/configuring-commit-squashing-for-pull-requests) so I don't have to rebase the individual commits on my fork?

matthew-carroll commented:
> I don't think squashing is currently enabled for this repo. Could you [enable it](https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/configuring-pull-request-merges/configuring-commit-squashing-for-pull-requests) so I don't have to rebase the individual commits on my fork?

That's a setting that doesn't apply until you merge this PR into the main repository. This repo shouldn't cause any restrictions with regard to what you're able to do with your own branch on your own machine. Let me know if there's a specific action you can't seem to execute.

matthew-carroll commented:
> I created a ShapeStyle class but maybe it would be better to just have color and gradient be properties on Rectangle. Also, ShapeStyle in SwiftUI doesn't have a color property. I think it might get color from the ForegroundStyle. It doesn't have a gradient property either. Instead it has a linearGradient, radialGradient, etc. That didn't seem very Flutter-like to add all of those properties when they can all be expressed with Gradient.

Unless specific hurdles come up, we should reflect the same API that exists in Swift UI. Are there any details that you're blocked from replicating in this case?

> Should there be some more general Shape class that ties all of the shapes together, or should we implement each shape widget independently?

Is there a corresponding Swift UI API that you're thinking of? If not, let's just stick to replicating the public Swift UI APIs. If at a later point we find an opportunity to de-duplicate a bunch of implementation details behind the scenes, then we can consider that.

suragch commented:
> That's a setting that doesn't apply until you merge this PR into the main repository. This repo shouldn't cause any restrictions with regard to what you're able to do with your own branch on your own machine.

Ah, I see. That's good to know. I was thinking that having multiple commits here would make your review harder by having view the commits separately. 

> Unless specific hurdles come up, we should reflect the same API that exists in Swift UI. Are there any details that you're blocked from replicating in this case?

In SwiftUI, `ShapeStyle` is a protocol that both `Color` and `Gradient` conform to. That means that the `fill` modifier can take either a `Color` or `Gradient`.

```swift
Rectangle()
    .fill(Color.green)
Rectangle()
    .fill(LinearGradient(gradient: Gradient(colors: [.yellow, .orange]), startPoint: .leading, endPoint: .trailing))
```

However, in Dart, `Color` and `Gradient` don't implement any other interface, so this isn't possible:

```dart
Rectangle(
  fill: Colors.blue,
),
Rectangle(
  fill: LinearGradient(colors: [Colors.yellow, Colors.orange]),
),
```

In my current PR I created a `ShapeStyle` class that could wrap a `Color` or `Gradient`:

```
Rectangle(
  fill: ShapeStyle(
    color: Colors.blue,
  ),
),
Rectangle(
  fill: ShapeStyle(
    gradient: LinearGradient(
      colors: [Colors.yellow, Colors.orange],
    ),
  ),
),
```

Is that the best way to do it given the constraints?

After thinking more about it, making `Color` an unnamed parameter would be better:

```
Rectangle(
  fill: ShapeStyle(Colors.blue),
),
```

> Is there a corresponding Swift UI API that you're thinking of? If not, let's just stick to replicating the public Swift UI APIs. If at a later point we find an opportunity to de-duplicate a bunch of implementation details behind the scenes, then we can consider that.

There is a [Shape](https://developer.apple.com/documentation/swiftui/shape) protocol with the following types that conform to it:

- [AnyShape](https://developer.apple.com/documentation/swiftui/anyshape)
- [ButtonBorderShape](https://developer.apple.com/documentation/swiftui/buttonbordershape)
- [Capsule](https://developer.apple.com/documentation/swiftui/capsule)
- [Circle](https://developer.apple.com/documentation/swiftui/circle)
- [ContainerRelativeShape](https://developer.apple.com/documentation/swiftui/containerrelativeshape)
- [Ellipse](https://developer.apple.com/documentation/swiftui/ellipse)
- [OffsetShape](https://developer.apple.com/documentation/swiftui/offsetshape)
- [Path](https://developer.apple.com/documentation/swiftui/path)
- [Rectangle](https://developer.apple.com/documentation/swiftui/rectangle)
- [RotatedShape](https://developer.apple.com/documentation/swiftui/rotatedshape)
- [RoundedRectangle](https://developer.apple.com/documentation/swiftui/roundedrectangle)
- [ScaledShape](https://developer.apple.com/documentation/swiftui/scaledshape)
- [TransformedShape](https://developer.apple.com/documentation/swiftui/transformedshape)
- [UnevenRoundedRectangle](https://developer.apple.com/documentation/swiftui/unevenroundedrectangle)

But I agree that we should just implement the concrete shapes themselves for now.

## Conclusion

Unless you have a new idea, I'm going to make `Color` an unnamed optional parameter of `ShapeStyle` and then continue in this general direction to finish the basic `Rectangle` implementation (i.e., `stroke` will also take a `ShapeStyle`).

suragch commented:
@matthew-carroll Ready for a final review.

The SwiftUI `Rectangle` examples are on the left and the Flutter ones are one the right:

<img width="206" alt="Screenshot 2024-06-08 at 10 59 08" src="https://github.com/Flutter-Bounty-Hunters/swift_ui/assets/10300220/b90a9f5e-8953-41ad-9f6f-95063a70bd78">
<img width="233" alt="Screenshot 2024-06-08 at 10 59 24" src="https://github.com/Flutter-Bounty-Hunters/swift_ui/assets/10300220/6ebf79a6-ba72-4183-8812-ddf57f0dc4dd">

I dropped the `cornerRadius` and `shadow` examples because these are view-level modifications in SwiftUI, not shape-level ones. In the future we'll need to create composable `CornerRadius` and `Shadow` widgets to implement those.

I mentioned my intention previously to have color be an optional unnamed parameter of ShapeStyle:

```
Rectangle(
  fill: ShapeStyle(Colors.blue),
),
```

However, it isn't possible in Dart to mix optional positional and named parameters. Upvote [this issue](https://github.com/dart-lang/language/issues/1076) if you think it should be.

So I'm reverting to the previous version:

```
Rectangle(
  fill: ShapeStyle(
    color: Colors.blue,
  ),
),
```

Originally I implemented `Rectangle` with a `Container` but the `BoxDecoration` didn't make gradient strokes very convenient so I replaced it with a `CustomPainter`.

matthew-carroll commented:
@suragch is there anything that the `ShapeStyle` is doing beyond just holding the color/gradient?

If not, is there any reason not to take a `color` or `gradient` directly within the `Rectangle` widget?

One option there would be to use two properties: `fillColor`, `fillGradient`.

I'm open to multiple approaches, but I wanted to check the rationale for `ShapeStyle`.

suragch commented:
@matthew-carroll The sole purpose for `ShapeStyle` is to attempt to match the type system in SwiftUI. It currently does nothiing more than wrap color and gradient. As a Flutter dev, I would much more prefer to use `fillColor` and `fillGradient` directly under `Rectangle` and just drop `ShapeStyle`. I think this would also be intuitive enough for SwiftUI devs.

matthew-carroll commented:
Sounds good. Let's do that.

suragch commented:
@matthew-carroll I removed `ShapeStyle` in favor of `fillColor`, `fillGradient`, `strokeColor`, and `strokeGradient` parameters on `Rectangle`.

matthew-carroll commented on .gitignore:
```diff
@@ -27,3 +27,6 @@ migrate_working_dir/
 build/
+
+# Don't check in golden failure output.
+**/failures/*.png
```
We probably want to ignore everything under a `/falures` directory, but only where goldens might live.

```
test/**/failures/**
test_goldens/**/failures/**
```

matthew-carroll commented on lib/src/layout_adjustments/frame.dart:
```diff
@@ -9,9 +10,22 @@ class Frame extends StatelessWidget {
+
+  /// The alignment of the child within the frame.
+  ///
+  /// See [Alignment] for more info about the available alignments.
```
Can you confirm that `Alignment` includes the same set of values as Swift UI's corresponding alignment that it exposes within `frame()`? For example, I ended up creating custom alignment enumerations for `VStack` and `HStack` because those alignments weren't the same as Flutter's.

suragch replied:
You're right. `Alignment` is different in Flutter than in SwiftUI. Should we solve this with an extension? Recreate another `Alignment` class? Create another alignment class with a different name?

Should we solve this in the current PR or start a new issue for that?

matthew-carroll replied:
We should solve it in this PR. I suggest creating an enum with the same name and values as Swift UI.

matthew-carroll commented on lib/src/shapes/rectangle.dart:
```diff
@@ -0,0 +1,160 @@
+
+  @override
+  void paint(Canvas canvas, Size size) {
+    _paintFill(canvas, size, fillColor, fillGradient);
```
You probably don't need to pass the painter properties to other methods, because those methods already have access to properties.

matthew-carroll commented on test/shapes/rectangle_test.dart:
```diff
@@ -0,0 +1,3 @@
+void main() {
+  // TODO:
```
Unless you plan to do this in this PR, please remove this file.

suragch commented:
@matthew-carroll You're suggestions are implemented now. Let me know if there are any further adjustments to make.

suragch commented on lib/src/layout_adjustments/alignment.dart:
```diff
@@ -0,0 +1,117 @@
+enum HorizontalAlignment {
```
`Alignment` in SwiftUI can be defined in terms of `HorizontalAlignment` and `VerticalAlignment` so that's why I moved them to a common file rather than importing `HStack` and `VStack` into the `Frame` file.

The `layout_adjustments` already existed and for the `Frame` widget, so that's why I put `alignment.dart` there. That's sounds good to move `alignment.dart` to the `layout` directory. I assume I should move `frame.dart` there too?

suragch commented on lib/src/layout_adjustments/alignment.dart:
```diff
@@ -0,0 +1,117 @@
+
+enum Alignment {
+  /// Alignment that's equidistant from all edges of the bounding space.
+  center(horizontal: HorizontalAlignment.center, vertical: VerticalAlignment.center),
```
That's a SwiftUI thing.

suragch commented on lib/src/layout_adjustments/frame.dart:
```diff
@@ -1,17 +1,33 @@
     this.width,
     this.height,
-    this.alignment = Alignment.center,
+    this.alignment = swift_ui.Alignment.center,
```
I was trying to do that but still needed `Alignment` from the widgets library to pass to the Align widget. I think it might work if I follow your other suggestion to make a `toFlutter` method.

suragch commented on lib/src/shapes/rectangle.dart:
```diff
@@ -0,0 +1,149 @@
+
+  @override
+  Widget build(BuildContext context) {
+    return SizedBox.expand(
```
I didn't realize there was a `size` property on `CustomPaint`. Thanks, I'll refactor with that.

suragch commented on lib/src/layout_adjustments/frame.dart:
```diff
@@ -50,9 +68,40 @@ class Frame extends StatelessWidget {
     );
   }
+
+  AlignmentGeometry _alignment(TextDirection direction) {
```
It would be possible to make the `TextDirection` default to `ltr` but Flutter itself tries hard not to do that, so I'm leaving it as a required parameter: `toFlutter(TextDirection textDirection)`.

suragch commented:
@matthew-carroll Assuming the `frame.dart` move is correct, I've applied the new set of suggestions.

matthew-carroll reviewed (approved):
LGTM - Thanks.
